[Feature] Add initializing timeout for RayService

seanlaii opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Currently, RayService will be stuck in Initializing indefinitely if something bad happens in the underlying pods, such as ImagePullBackOff.
To avoid wasting resources for the failed RayCluster, we decided to provide a way for user to configure the initializing timeout for RayService. 
Users can now configure the initializing timeout via annotation `ray.io/initializing-timeout`. When the RayService reaches the timeout, controller will mark its `RayServiceReady` condition with reason `InitializingTimeout`. Once the RayService is in this condition, controller will not create new cluster for the RayService, unless the controller detects update events from the RayService (`Generation` change).
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->
Closes #4138

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
  - [x] Unit tests
  - [x] Manual tests
  - [ ] This PR is not tested :(

## Manual Testing Result
1. Create a RayService by adding `ray.io/initializing-timeout: "10"` annotation in [ray-service.sample.yaml](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-service.sample.yaml).
```zsh
$ k get rayservice
NAME                SERVICE STATUS   NUM SERVE ENDPOINTS
rayservice-sample
```
2. Since the RayCluster is not ready in 10 seconds, the controller detects the timeout and emits:
```zsh
{"level":"info","ts":"2025-10-25T04:50:09.953Z","logger":"controllers.RayService","msg":"RayService initializing timeout exceeded","RayService":{"name":"rayservice-sample","namespace":"default"},"reconcileID":"5a565e9c-d6b1-4c9f-9eb6-0aca36a24860","timeout":10,"timeInInitializing":10.953153549,"generation":1}
```
3. Since the RayCluster name is removed from the RayService status, the RayCluster is scheduled to be removed:
```zsh
{"level":"info","ts":"2025-10-25T04:50:11.964Z","logger":"controllers.RayService","msg":"Scheduled dangling RayCluster for deletion","RayService":{"name":"rayservice-sample","namespace":"default"},"reconcileID":"a6fd1eef-1227-4565-98b7-734d3f3ba8e5","rayClusterName":"rayservice-sample-pts9w","deletionDelay":"1m0s","deletionTimestamp":"2025-10-25T04:51:11.964Z"}
```
4. After reaching the deletionTimestamp, the RayCluster is deleted:
```
{"level":"info","ts":"2025-10-25T04:51:12.093Z","logger":"controllers.RayService","msg":"reconcileRayCluster","RayService":{"name":"rayservice-sample","namespace":"default"},"reconcileID":"5846ef6c-e5d3-4c0d-9a82-4de7a0af2145","delete Ray cluster":"rayservice-sample-pts9w","reason":"Deletion timestamp 2025-10-25 04:51:11.964204675 +0000 UTC m=+149.210051208 for RayCluster rayservice-sample-pts9w has passed. Deleting cluster immediately."}
```

rueian commented on ray-operator/controllers/ray/rayservice_controller.go:
I think we should also make some changes around here.

seanlaii commented on ray-operator/controllers/ray/rayservice_controller.go:
```diff
@@ -306,6 +314,19 @@ func (r *RayServiceReconciler) calculateStatus(ctx context.Context, rayServiceIn
+		if readyCond.ObservedGeneration < rayServiceInstance.Generation {
+			// New generation - clear the timeout condition and let logic below recalculate
+			rayServiceInstance.Status.Conditions = []metav1.Condition{}
+		} else {
```
Got it. So even if they update the RayService to a correct config, we still don't recreate the RayCluster. Is my understanding correct?

rueian replied:
That's correct.  They need to delete the original one first.

rueian commented:
Hi @seanlaii, please help resolve the conflict!

seanlaii commented:
> Hi @seanlaii, please help resolve the conflict!

Hi @rueian , thanks for the info! I have addressed the above comments. Please help take a look when you get a chance. Thanks!

win5923 reviewed (commented):
Nice! I’ve tested it locally, and it works well.
```
$ kubectl describe rayservice rayservice-sample
Name:         rayservice-sample
Namespace:    default
...
Status:
  Active Service Status:
    Ray Cluster Status:
      Desired CPU:     0
      Desired GPU:     0
      Desired Memory:  0
      Desired TPU:     0
      Head:
  Conditions:
    Last Transition Time:  2025-10-26T10:11:33Z
    Message:               RayService failed to become ready within the configured timeout of 10s. Time spent initializing: 12.157851061s
    Observed Generation:   1
    Reason:                InitializingTimeout
    Status:                False
    Type:                  Ready
    Last Transition Time:  2025-10-26T10:11:33Z
    Message:               RayService is initializing
    Observed Generation:   1
    Reason:                Initializing
    Status:                False
    Type:                  UpgradeInProgress
  Last Update Time:        2025-10-26T10:11:45Z
```

```
$ k get raycluster
No resources found in default namespace.
```
